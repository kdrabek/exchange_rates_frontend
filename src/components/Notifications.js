import React from 'react';
import { Column, Row, Switch } from 'react-foundation';

class Notifications extends React.Component {
  render() {
    return (
      <Row>
        <Column large={3}>
          <h4 classNameName="subheader">Jak to działa?</h4>
          <hr/>
          <ul>
            <li><p>Notifikacje wysyłane są <em>raz dziennie</em>, po ściągnięciu aktualnych kursów walut.</p></li>
            <li><p>Notyfikacja zostanie wysłana tylko wtedy, jeśli wybrany warunek jest spełniony (kurs powyżej lub poniżej progu).</p></li>
            <li><p>Tylko <em>aktywne</em> notyfikacje będą wysyłane.</p></li>
          </ul>
        </Column>
        <Column large={9}>
          <h4 className="subheader">Moje notyfikacje</h4>
          <hr/>
          <table className="unstriped hover">
            <thead>
              <tr>
                <th width="300">Waluta</th>
                <th width="150">Próg</th>
                <th width="150">Warunek</th>
                <th width="150">Aktywna?</th>
                <th width="150"></th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <select disabled>
                    <option value="husker" selected>AUD</option>
                    <option value="starbuck">Starbuck</option>
                    <option value="hotdog">Hot Dog</option>
                    <option value="apollo">Apollo</option>
                  </select>
                </td>
                <td>
                  <input type="number" step="0.01" min="0" value="2.35" disabled />
                </td>
                <td>
                  <select disabled>
                    <option value="husker" selected>powyżej</option>
                    <option value="starbuck">poniżej</option>
                  </select>
                </td>
                <td>
                  <Switch />
                </td>
                <td>
                  <button type="button" className="warning button">Usuń</button>
                </td>
              </tr>

              <tr>
                <td>
                  <select disabled>
                    <option value="husker" selected>USD</option>
                    <option value="starbuck">Starbuck</option>
                    <option value="hotdog">Hot Dog</option>
                    <option value="apollo">Apollo</option>
                  </select>
                </td>
                <td>
                  <input type="number" step="0.01" min="0" value="3.56" disabled />
                </td>
                <td>
                  <select disabled>
                    <option value="husker">powyżej</option>
                    <option value="starbuck" selected>poniżej</option>
                  </select>
                </td>
                <td>
                  <Switch />
                </td>
                <td>
                  <button type="button" className="warning button">Usuń</button>
                </td>
              </tr>

              <tr>
                <td>
                  <select>
                    <option value="husker" selected></option>
                    <option value="starbuck">AUD</option>
                    <option value="hotdog">CHF</option>
                    <option value="apollo">HUF</option>
                  </select>
                </td>
                <td>
                  <input type="number" step="0.01" min="0" value="0.00" />
                </td>
                <td>
                  <select>
                    <option value="husker">powyżej</option>
                    <option value="starbuck">poniżej</option>
                  </select>
                </td>
                <td>

                </td>
                <td>
                  <button type="button" className="success button">Dodaj</button>
                </td>
              </tr>
            </tbody>
          </table>

        </Column>
      </Row>
    );
  }
}

export default Notifications;
