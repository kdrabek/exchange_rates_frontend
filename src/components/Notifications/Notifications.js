import React, { Component } from 'react';
import { Row, Col, Table, FormControl, Button, FormGroup, ControlLabel, Form } from 'react-bootstrap';
import Switch from 'react-bootstrap-switch';
import { connect } from 'react-redux';

import * as notificationsActions from '../../actions/notificationsActions';
import * as currenciesActions from '../../actions/currenciesActions';

const thresholdConditionMap = {
  'ABOVE': 'powyżej', 'BELOW': 'poniżej'
};

class Notifications extends Component {

  constructor(props) {
    super(props);
    this.submitForm = this.submitForm.bind(this);
  }
  
  componentDidMount() {
    const authToken = localStorage.getItem('AuthUserToken');
    this.props.loadNotifications(authToken);
    this.props.loadCurrencies(authToken);
  }
  
  deleteNotification(notificationId) {
    const authToken = localStorage.getItem('AuthUserToken');
    this.props.deleteNotification(authToken, notificationId);    
  }

  changeNotificationState(notification){
    const authToken = localStorage.getItem('AuthUserToken');
    notification.is_active = !notification.is_active;
    this.props.updateNotification(authToken, notification);
  }

  displayNotifications(){
    let notificationsTable = this.props.notifications.map((notification, index) => {
      return (
        <tr key={index}>
          <td>
            <FormControl componentClass="select" placeholder="select" disabled>
              <option value="select" selected>{notification.currency}</option>
            </FormControl>
          </td>
          <td>
            <FormControl type="number" step="0.01" min="0" value={notification.rate} disabled />
          </td>
          <td>
            <FormControl componentClass="select" placeholder="select" disabled>
              <option value="select" selected>{thresholdConditionMap[notification.threshold]}</option>
            </FormControl>
          </td>
          <td>
            <Switch
              value={notification.is_active}
              onColor="success"
              offColor="warning"
              onChange={this.changeNotificationState.bind(this, notification)}
            />
          </td>
          <td>
            <Button 
              bsStyle="warning"
              onClick={this.deleteNotification.bind(this, notification.id)}
            >Usuń</Button>
          </td>
        </tr>
      );
    });
    return notificationsTable;
  }

  submitForm(e){
    e.preventDefault();
    const authToken = localStorage.getItem('AuthUserToken');
    const newNotification = {
      code: e.target.currency.value,
      rate: parseFloat(e.target.rate.value),
      threshold: e.target.threshold.value,
      is_active: true
    };
    this.props.addNotification(authToken, newNotification);
  }

  displayAddNotificationForm() {
    const currencySelect = this.props.currencies.length ? this.props.currencies.map(currency => {
      return <option
        value={currency.code}>{currency.code} ({currency.country})</option>
      }
    ) : <option></option>;
    return (
      <Form onSubmit={this.submitForm}>
        <FormGroup>
          <ControlLabel>Waluta</ControlLabel>
          <FormControl componentClass="select" placeholder="select" name="currency">
            {currencySelect}
          </FormControl>
        </FormGroup>

        <FormGroup>
          <ControlLabel>Próg</ControlLabel>
          <FormControl type="number" step="0.01" min="0" placeholder="0.00" name="rate" required/>
        </FormGroup>

        <FormGroup>
          <ControlLabel>Warunek</ControlLabel>
          <FormControl componentClass="select" placeholder="select" name="threshold">
            <option value="ABOVE" selected>powyżej</option>
            <option value="BELOW">poniżej</option>
          </FormControl>
        </FormGroup>

        <FormGroup>
          <Button type="submit" bsStyle="success" >Dodaj</Button>
        </FormGroup>
      </Form>
    )
  }
  
  render() {
    const notificationsTable = this.props.notifications ? this.displayNotifications() : 'Sciagam...';
    const notificationsForm = this.displayAddNotificationForm();
    return (
      <div>
        <Row>
          <Col md={3}>
            <h4>Jak to działa?</h4>
            <hr/>
            <ul>
              <li><p>Notyfikacja zostanie wysłana tylko wtedy, jeśli wybrany warunek jest spełniony (kurs powyżej lub poniżej progu).</p></li>
              <li><p>Tylko <em>aktywne</em> notyfikacje będą wysyłane.</p></li>
            </ul>
          </Col>
          <Col md={9}>
            <h4>Moje notyfikacje</h4>
            <hr/>
            <Table striped hover>
              <thead>
                <tr>
                  <th>Waluta</th>
                  <th>Próg</th>
                  <th>Warunek</th>
                  <th>Aktywna</th>
                </tr>
              </thead>
              <tbody>
                {notificationsTable}
              </tbody>
            </Table>
          </Col>
        </Row>
        <Row>
          <Col md={4} mdOffset={3}>
            <h4>Dodaj nową notyfikację</h4>
            {notificationsForm}
          </Col>
        </Row>
      </div>  
    );
    
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    notifications: state.notifications,
    currencies: state.currencies
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    loadNotifications: (token) => dispatch(notificationsActions.loadNotifications(token)),
    addNotification: (token, notification) => dispatch(notificationsActions.addNotification(token, notification)),
    updateNotification: (token, notification) => dispatch(notificationsActions.updateNotification(token, notification)),
    deleteNotification: (token, id) => dispatch(notificationsActions.deleteNotification(token, id)),
    loadCurrencies: (token) => dispatch(currenciesActions.loadCurrencies(token))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Notifications);
