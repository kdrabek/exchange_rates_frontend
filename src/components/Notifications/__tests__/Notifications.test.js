import React from 'react';
import ReactDOM from 'react-dom';
import { Notifications } from '../Notifications';
import renderer from 'react-test-renderer';

class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key];
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    this.store[key] = undefined;
  }
};

global.localStorage = new LocalStorageMock;

describe('Notifications component', () => {

  it('matches the snapshot', () => {
    const currencies = [{
      code: 'ABC', name: 'awesome currency', country: 'country', table_type: 'A'
    }];
    const notifications = [];
    
    const component = renderer.create(
      <Notifications 
        currencies={currencies} 
        notifications={notifications} 
        loadNotifications={jest.fn()}
        addNotification={jest.fn()}
        updateNotification={jest.fn()}
        deleteNotification={jest.fn()}
        loadCurrencies={jest.fn()}
      />
    );
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

});
