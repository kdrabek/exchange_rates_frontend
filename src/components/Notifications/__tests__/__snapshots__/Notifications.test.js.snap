exports[`Notifications component matches the snapshot 1`] = `
<div>
  <div
    className="row">
    <div
      className="col-md-3">
      <h4>
        Jak to działa?
      </h4>
      <hr />
      <ul>
        <li>
          <p>
            Notyfikacja zostanie wysłana tylko wtedy, jeśli wybrany warunek jest spełniony (kurs powyżej lub poniżej progu).
          </p>
        </li>
        <li>
          <p>
            Tylko 
            <em>
              aktywne
            </em>
             notyfikacje będą wysyłane.
          </p>
        </li>
      </ul>
    </div>
    <div
      className="col-md-9">
      <h4>
        Moje notyfikacje
      </h4>
      <hr />
      <table
        className="table table-striped table-hover">
        <thead>
          <tr>
            <th>
              Waluta
            </th>
            <th>
              Próg
            </th>
            <th>
              Warunek
            </th>
            <th>
              Aktywna
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
  <div
    className="row">
    <div
      className="col-md-4 col-md-offset-3">
      <h4>
        Dodaj nową notyfikację
      </h4>
      <div>
        <form
          className=""
          onSubmit={[Function]}>
          <div
            className="form-group">
            <label
              className="control-label"
              htmlFor={undefined}>
              Waluta
            </label>
            <select
              className="form-control"
              id={undefined}
              name="currency"
              placeholder="select"
              type={undefined}>
              <option
                value="ABC">
                ABC
                 (
                country
                )
              </option>
            </select>
          </div>
          <div
            className="form-group">
            <label
              className="control-label"
              htmlFor={undefined}>
              Próg
            </label>
            <input
              className="form-control"
              id={undefined}
              min="0"
              name="rate"
              placeholder="0.00"
              required={true}
              step="0.01"
              type="number" />
          </div>
          <div
            className="form-group">
            <label
              className="control-label"
              htmlFor={undefined}>
              Warunek
            </label>
            <select
              className="form-control"
              id={undefined}
              name="threshold"
              placeholder="select"
              type={undefined}>
              <option
                selected={true}
                value="ABOVE">
                powyżej
              </option>
              <option
                value="BELOW">
                poniżej
              </option>
            </select>
          </div>
          <div
            className="form-group">
            <button
              className="btn btn-success"
              disabled={false}
              type="submit">
              Dodaj
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
